apps/test/unit/code-studio/pd/form_components_func/FormControllerTest.js

```diff
@@ -221,13 +221,14 @@ describe('FormController', () => {
         expect(getData(DummyPage1)).to.eql({...testData});
       });
 
-      it('Disables the save button during save', () => {
+      it('Disables the save button during save and renders spinner', () => {
         form = isolateComponent(<FormController {...defaultProps} />);
         form.findAll('Button')[1].props.onClick();
         expect(form.findAll('Button')[1].props.disabled).to.be.true;
+        expect(form.findAll('Spinner')).to.have.length(1);
       });
 
-      it('Re-enables the save button after successful save', () => {
+      it('Re-enables the save button after successful save and removes spinner', () => {
```

Comment:
what happens if the request fails? you can figure this out by using the request blocking feature in Chrome: 

<img width="639" alt="Screen Shot 2022-03-07 at 2 14 31 PM" src="https://user-images.githubusercontent.com/8001765/157118951-b63ccbfb-059a-4220-9cbf-fdd976e42bb6.png">

Reply:
I matched behavior of what happens if submit fails, so this function is called on both failures:
```
const handleRequestFailure = data => {
    if (data?.responseJSON?.errors?.form_data) {
      setErrors(data.responseJSON.errors.form_data);
      setErrorHeader(i18n.formErrorsBelow());
    } else {
      // Otherwise, something unknown went wrong on the server
      setGlobalError(true);
      setErrorHeader(i18n.formServerError());
    }
    setSubmitting(false);
    setSaving(false);
  };
```
I didn't add tests specifically for this for saving. Thoughts?

Reply:
I also haven't "seen" this happen in the browser––will explore tomorrow!

Reply:
this behavior sounds good. a test case like these two, which verifies what happens when you get a 5xx response with no responseJSON, would be great.

Reply:
I'm realizing both submit and save have a test for an error that's still a 400 ... will refactor separately and also include 500 responses in the testing.